File: LLD/cpp/ObserverPattern/Observable/IphoneObservable/IphoneObservable.cpp
// #include "ObserverPattern/Observable/IphoneObservable.h"
#include "IphoneObservable.h"
// #include "Observer/NotificationAlertObserver/NotificationAlertObserver.h" // Include necessary headers

IphoneObservable::IphoneObservable() : stockCount(0) {}

void IphoneObservable::add(NotificationAlertObserver* observer) {
    observerList.push_back(observer);
}

void IphoneObservable::remove(NotificationAlertObserver* observer) {
    observerList.remove(observer);
}

void IphoneObservable::notifySubscriber() {
    for (NotificationAlertObserver* observer : observerList) {
        observer->update();
    }
}

void IphoneObservable::setStockCount(int newStockAdded) {
    if (stockCount == 0) {
        notifySubscriber();
    } 
    stockCount += newStockAdded;
}

int IphoneObservable::getStockCount() {
    return stockCount;
}

File: LLD/typescript/ObserverPattern/Observable/IphoneObservable.ts
import NotificationAlertObserver from "../Observer/NotificationAlertObserver";
import StocksObservable from "./StocksObservable";
class IphoneObservable implements StocksObservable {
    private observers: NotificationAlertObserver[] = [];
    private stockCount: number = 0;

    add(observer: NotificationAlertObserver): void {
        this.observers.push(observer);
    }

    remove(observer: NotificationAlertObserver): void {
        const index = this.observers.indexOf(observer);
        if (index !== -1) {
            this.observers.splice(index, 1);
        }
    }

    notifySubscribers(): void {
        for (const observer of this.observers) {
            observer.update();
        }
    }

    setStockCount(newStockAdded: number): void {
        if (this.stockCount === 0) {
            this.notifySubscribers();
        }
        this.stockCount += newStockAdded;
    }

    getStockCount(): number {
        return this.stockCount;
    }
}
export default IphoneObservable;